### PR TITLE
Handle profile endpoint in Gateway outbound stack

### PR DESCRIPTION
**Pull request by mateiidavid**
## What

When testing [linkerd2/PR#6090](https://github.com/linkerd/linkerd2/pull/6090) we noticed that requests to a mirror service from a source cluster to a target cluster would fail on the outbound side of the gateway (in the target cluster) if the profile returned contains an endpoint. [[1]](https://github.com/linkerd/linkerd2/pull/6090#issuecomment-871754124). The failure is related to some changes that we have made previously to support StatefulSets across multicluster ([PR#6260 in linkerd2](https://github.com/linkerd/linkerd2/pull/6260)); when we send a request to an endpoint mirror service (a mirror service that corresponds to a pod), we expect the proxy to connect to it directly instead of building a logical stack.

Currently, the outbound proxy does not handle endpoints when running in gateway mode. This change is focused on adding support for direct endpoint communication; when a profile contains an endpoint, we will establish a connection directly to the endpoint and avoid building a load balancer.

## How

The change adds two `push_switch` layers: one for the TCP stack and one for the HTTP stack.

### TCP stack

* Replaced the request filter with a push switch. Before, the request filter would fail if the resolved profile would not have a logical address or if no profile would be discovered for the target. Now, we handle the `no discovery` case in the `push_switch`, however, we no longer error out when the profile does not contain a logical address.

* Created a logical and an endpoint stack. If the profile contains an endpoint, then we use the endpoint stack. If it contains a logical address, we use the logical stack instead.

* Changed the visibility for `Endpoint::from_metadata()` and `Outbound::no_tls_reason()` so we can construct the endpoints more easily (in both the TCP and HTTP stacks).

### HTTP stack

* Largely the same concept as above; to make it easier to build the endpoints, we're passing the `no_tls_message` through to the `NewGateway` type. 

* Changed the trait bounds for the `NewService` impl on the gateway to be an `Either`. When we create the new gateway,  if the profile has an endpoint, we return an `Either::B::<HttpEndpoint>` otherwise we return an `Either::A::<HttpLogical`. In the switch layer, we simply return the service as a result; this is because the compiler couldn't refer the return type, making it a result explicitly seemed to help.

Lastly, changed the unit tests a bit, first to accommodate the new constructor changes for `NewGateway` and then to add a new case for the endpoint stack.

### Manual testing
---

To see the changes play out in an actual multicluster scenario, I deployed linkerd in two clusters using the [PR branch build](https://github.com/linkerd/linkerd2/pull/6090). I applied two manifests to the target cluster: an nginx statefulset and an nginx deployment (manifests for the statefulset are taken from Kevin's comment [here](https://github.com/linkerd/linkerd2/pull/6090#pullrequestreview-695516895)). I then mirrored the statefulset service and the deployment service in the source cluster and also created a curl pod to send requests to the mirror services.

Without the proxy changes, I expected the following scenarios:
  1. For `nginx-deployment`: exec onto curl pod, send a request to the mirror service (nginx deployment) and look at the gateway logs in the target cluster to see if the request was redirected on the outbound side to its exported service. Since the deployment service is _not_ headless, we expect this to **succeed* (it has a logical addr in the response).
  2. For `nginx-statefulset`: exec onto curl pod, send a request to an endpoint mirror service (a clusterIP service for a specific pod in the target cluster). In the logs, we expect this to **fail** with a 404 status code and a `BadDomain` error since the profile will contain an endpoint in the response.

Logs:
```
# services, pods & control plane version
# east = source cluster, west = target cluster

❯ k get svc --context=k3d-east
NAME                    TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
kubernetes              ClusterIP   10.43.0.1       <none>        443/TCP   22h
nginx-svc-west          ClusterIP   None            <none>        80/TCP    22h
nginx-deploy-svc-west   ClusterIP   10.43.155.254   <none>        80/TCP    21h
nginx-ss-0-west         ClusterIP   10.43.145.231   <none>        80/TCP    15m
nginx-ss-1-west         ClusterIP   10.43.138.0     <none>        80/TCP    15m
nginx-ss-2-west         ClusterIP   10.43.238.163   <none>        80/TCP    15m

❯ k get po --context=k3d-east
NAME                    READY   STATUS    RESTARTS   AGE
curl-56dc7d945d-qv7dd   2/2     Running   0          22h

❯ bin/linkerd version
Client version: git-5bfe4252
Server version: git-5bfe4252

-----


# exec curl pod
$ curl  http://nginx-deploy-svc-west.default.svc.east.cluster.local:80

# expect it to work
[   211.010516s] TRACE ThreadId(01) inbound:accept{client.addr=10.42.0.1:54684}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}: linkerd_timeout::failfast: HTTP Logical has become ready
[   211.010565s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:54684}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.19:80}:orig-proto-upgrade: linkerd_proxy_http::client: method=GET uri=http://nginx-deploy-svc.default.svc.west.cluster.local:80/ version=HTTP/1.1
[   211.010598s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:54684}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.19:80}:orig-proto-upgrade: linkerd_proxy_http::client: headers={"accept": "*/*", "user-agent": "curl/7.77.0-DEV", "l5d-client-id": "default.default.serviceaccount.identity.linkerd.east.cluster.local", "forwarded": "by=linkerd-gateway.linkerd-multicluster.serviceaccount.identity.linkerd.west.cluster.local;for=default.default.serviceaccount.identity.linkerd.east.cluster.local;host=nginx-deploy-svc.default.svc.west.cluster.local;proto=https", "host": "nginx-deploy-svc.default.svc.west.cluster.local", "l5d-dst-canonical": "nginx-deploy-svc.default.svc.west.cluster.local:80"}
[   211.010655s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:54684}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.19:80}:orig-proto-upgrade: linkerd_proxy_http::orig_proto: Upgrading request version=HTTP/1.1 absolute_form=false
[   211.012597s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:54684}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.20:80}:h2: linkerd_proxy_transport::metrics: client connection open
[   211.014500s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:54684}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.21:80}:h2: linkerd_proxy_transport::metrics: client connection open
[   211.014931s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:54684}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.20:80}: linkerd_reconnect: Connected
[   211.014977s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:54684}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.21:80}: linkerd_reconnect: Connected

---

$ curl http://nginx-ss-0.nginx-svc-west.default.svc.east.cluster.local:80

# expect it to fail
[   411.043980s] TRACE ThreadId(01) inbound:accept{client.addr=10.42.0.1:64673}:server{port=4143}:direct:gateway{dst=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80}:http{v=h2}: linkerd_timeout::failfast: Gateway has become ready
[   411.044186s]  WARN ThreadId(01) inbound:accept{client.addr=10.42.0.1:64673}:server{port=4143}:direct:gateway{dst=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80}:http{v=h2}: linkerd_app_core::errors: Failed to proxy request: bad domain client.addr=10.42.0.1:64673
[   411.044216s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:64673}:server{port=4143}:direct:gateway{dst=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80}:http{v=h2}: linkerd_app_core::errors: Closing server-side connection
[   411.044229s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:64673}:server{port=4143}:direct:gateway{dst=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80}:http{v=h2}: linkerd_app_core::errors: Handling error with HTTP response status=404 Not Found version=HTTP/2.0
```

With the proxy changes, we expect both to succeed:

```
# proxy version tagged as dev:
$ k get po linkerd-gateway-58c968f4f8-7cgmr -o yaml | rg 'image'
    image: cr.l5d.io/linkerd/proxy:dev
    ... 

# nginx-deployment
 $ curl http://nginx-deploy-svc-west.default.svc.east.cluster.local:80
[   106.607344s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:46152}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.20:80}: linkerd_reconnect: Connected
[   106.607396s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:46152}:server{port=4143}:direct:gateway{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-deploy-svc.default.svc.west.cluster.local:80 v=1.x}:logical{dst=nginx-deploy-svc.default.svc.west.cluster.local:80}:concrete{addr=nginx-deploy-svc.default.svc.west.cluster.local:80}:endpoint{server.addr=10.42.0.21:80}: linkerd_reconnect: Connected


# nginx statefulset instance
$ curl http://nginx-ss-0.nginx-svc-west.default.svc.east.cluster.local:80
[   122.744939s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:6030}:server{port=4143}:direct:gateway{dst=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80 v=1.x}: linkerd_app_gateway::gateway: Passing request to outbound
[   122.744971s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:6030}:server{port=4143}:direct:gateway{dst=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80 v=1.x}:orig-proto-upgrade: linkerd_proxy_http::client: method=GET uri=http://nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80/ version=HTTP/1.1
[   122.744992s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:6030}:server{port=4143}:direct:gateway{dst=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80 v=1.x}:orig-proto-upgrade: linkerd_proxy_http::client: headers={"accept": "*/*", "user-agent": "curl/7.77.0-DEV", "l5d-client-id": "default.default.serviceaccount.identity.linkerd.east.cluster.local", "forwarded": "by=linkerd-gateway.linkerd-multicluster.serviceaccount.identity.linkerd.west.cluster.local;for=default.default.serviceaccount.identity.linkerd.east.cluster.local;host=nginx-ss-0.nginx-svc.default.svc.west.cluster.local;proto=https", "host": "nginx-ss-0.nginx-svc.default.svc.west.cluster.local"}
[   122.745016s] DEBUG ThreadId(01) inbound:accept{client.addr=10.42.0.1:6030}:server{port=4143}:direct:gateway{dst=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80}:http{v=h2}:gateway{target=nginx-ss-0.nginx-svc.default.svc.west.cluster.local:80 v=1.x}:orig-proto-upgrade: linkerd_proxy_http::orig_proto: Upgrading request version=HTTP/1.1 absolute_form=false
```

The assumptions I had around the tests seem to hold: when we were supposed to fail, we did and with the changes the request gets passed through. This is further verified by the responses we receive, without the changes we get no response from curl, with the changes, we get the nginx status page.

Signed-off-by: Matei David <matei@buoyant.io>

### CODE REVIEW

File: linkerd/app/gateway/src/tests.rs
Comment (olix0r) [inline linkerd/app/gateway/src/tests.rs:167]: Clippy is complaining about this. Since this is `&'static str` you can just `Copy` the reference without cloning:

```suggestion
        let suffix = self.suffix;
```

File: linkerd/app/gateway/src/tests.rs
Comment (olix0r) [inline linkerd/app/gateway/src/tests.rs:181]:
```suggestion
    async fn run_default_profile(self) -> Result<http::Response<http::BoxBody>, Error> {
        let allow = NameMatch::new(Some(dns::Suffix::from_str(self.suffix).unwrap()));
        let profile = if allow.matches(self.target.name()) {
            Some(support::profile::only(profiles::Profile {
                addr: Some(self.target.clone().into()),
                ..profiles::Profile::default()
            }))
        } else {
            None
        };

        let resp = self.run(profile).await;
        resp
    }
```

Clippy complains about the `suffix.clone()` -- suffix is just a `'static str` reference so it can be copied without cloning. Furthermore, it looks like the first `target.clone()` isn't needed since it's already cloned below.

File: linkerd/app/gateway/src/lib.rs
Comment (olix0r) [inline linkerd/app/gateway/src/lib.rs:193]: It turns out this can be simplified quite a bit as:

```suggestion
        .push_switch(Ok::<_, Never>, endpoint.into_stack())
```

First, I tried to remove the return signature by instead encoding the error type as `Ok::<_, Never>(t)`. (It's fine to use `Infallible` in general, but we already have `Never` in scope -- we could probably get rid of our `Never` type in favor of `Infallible` but that's a separate change). We can know the success value based on `t` -- but since the error variant is never constructed we need to explicitly annotate it.

Then, clippy gives me the warning:
```
error: redundant closure
   --> linkerd/app/gateway/src/lib.rs:188:13
    |
188 |             |t: svc::Either<outbound::http::Logical, outbound::http::Endpoint>| Ok::<_, Never>(t),
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `Ok::<_, Never>`
    |
note: the lint level is defined here
```

Which is saying that `Ok::<_, Never>` is equivalent to 
```rust
|t: svc::Either<outbound::http::Logical, outbound::http::Endpoint>| Ok::<_, Never>(t)
```

File: linkerd/app/gateway/src/lib.rs
Comment (olix0r) [inline linkerd/app/gateway/src/lib.rs:30]:
```suggestion
    convert::{TryFrom, TryInto},
```

File: linkerd/app/gateway/src/gateway.rs
Comment (olix0r) [inline linkerd/app/gateway/src/gateway.rs:91]: I don't think we actually need to wire `no_tls_reason` through the constructor. `no_tls_reason` is really to handle the case when TLS is disabled on a workload so we don't report "not provided by service discovery" when we should report "identity disabled".

But identity **can't** be disabled on the gateway. We require ALPN for protocol negotiation, etc. Just above this we enforce this:
https://github.com/linkerd/linkerd2-proxy/blob/5ef0c35ccea0719b322bffe9fb668dc44b049073/linkerd/app/gateway/src/gateway.rs#L72-L75

So, we should probably omit `no_tls_reason` on the gateway and just hardcode the NotProvidedByServiceDiscovery reason here.

File: linkerd/app/outbound/src/lib.rs
Comment (olix0r) [inline linkerd/app/outbound/src/lib.rs:96]: Once we remove `no_tls_reason` from the gateway, this can be private again
```suggestion
    fn no_tls_reason(&self) -> tls::NoClientTls {
```

File: linkerd/app/gateway/src/gateway.rs
Comment (mateiidavid) [reply linkerd/app/gateway/src/gateway.rs]: This makes sense. I was wondering whether we'd have a `None` value at any point or if we have a default value for this (so we can avoid making `no_tls_reason` public and pass it through). It's demystified now!

File: linkerd/app/gateway/src/lib.rs
Comment (mateiidavid) [reply linkerd/app/gateway/src/lib.rs]: This looks much cleaner, made the change. I went ahead and used `Never` as an error type here since it was already exported. I wasn't sure if I should pull in some of the changes from #1158 or wait until it is merged and then rebase off the latest main. Let me know what would work better here.

Comment (olix0r) [reply linkerd/app/gateway/src/lib.rs]: I wouldn't merge it in until it's on main. I was going to try to let this merge first to reduce overhead for you, or if the other gets merged first I'd probably merge main into this branch for you

**Review by olix0r — APPROVED** [review]
Great work! Change looks good to me. Welcome to proxyland.

File: linkerd/app/gateway/src/lib.rs
Comment (mateiidavid) [reply linkerd/app/gateway/src/lib.rs]: Cleaned it up.

**Review by hawkw — APPROVED** [review]
looks great!

since this is your first proxy PR, i thought you might appreciate a couple Rust style suggestions. none of these are blockers and i'd be very happy to see this merge as-is, but if you want to make additional changes in a follow-up, i mentioned a couple ways we could make the code more idiomatic.

great work!

**Comment by olix0r** [comment]
I made both of @hawkw's suggested changes. I'll merge when this passes CI

**Pull request merged**
